src/static/js/Changeset.js

```diff
@@ -968,18 +976,24 @@ class TextLinesMutator {
         this._curLine += newLines.length;
         // insert the remaining chars from the "old" line (e.g. the line we were in
         // when we started to insert new lines)
-        this._curSplice.push(theLine.substring(lineCol));
-        this._curCol = 0; // TODO(doc) why is this not set to the length of last line?
+        const remaining = theLine.substring(lineCol);
+        if (remaining !== '') this._curSplice.push(remaining);
+        this._curCol = 0;
       } else {
         this._curSplice.push(...newLines);
         this._curLine += newLines.length;
       }
+    } else if (!this.hasMore()) {
+      // There are no additional lines. Although the line is put into splice, curLine is not
+      // increased because there may be more chars in the line (newline is not reached). We are
+      // inserting at the end of lines. curCol is 0 as curLine is not in splice.
+      this._curSplice.push(text);
```

Comment:
I don't think this is right. If my understanding is correct, each entry in `this._lines` represents a line of the document and must end with a newline character. By pushing to `this._curSplice`, this code creates a new line at the end of the document, one that line doesn't end with a newline. This code should instead insert the text after column `this._curCol` (which should be before the line's newline) in line `this._curLine`.

I think the actual bug is inside `this._putCurLineInSplice()`: That method should not attempt to read past the end of `this._lines` or delete lines that don't exist in `this._lines`. Something like this (untested):

```diff
diff --git a/src/static/js/Changeset.js b/src/static/js/Changeset.js
index 5e0bc34fc..930a5e187 100644
--- a/src/static/js/Changeset.js
+++ b/src/static/js/Changeset.js
@@ -823,11 +823,18 @@ class TextLinesMutator {
    * @returns {number} the index of the added line in curSplice
    */
   _putCurLineInSplice() {
-    if (!this._isCurLineInSplice()) {
-      this._curSplice.push(this._linesGet(this._curSplice[0] + this._curSplice[1]));
-      this._curSplice[1]++;
+    assert(this._inSplice, 'not in splice');
+    assert(this._curLine >= this._curSplice[0], 'curLine is before splice start');
+    while (this._curSplice[0] + this._curSplice.length - 2 < this._curLine + 1) {
+      const n = this._curSplice[0] + this._curSplice[1];
+      if (n < this._linesLength()) {
+        this._curSplice.push(this._linesGet(n));
+        this._curSplice[1]++;
+      } else {
+        // No more lines available to add to the splice. Create a new line from scratch.
+        this._curSplice.push('\n');
+      }
     }
-    // TODO should be the same as this._curSplice.length - 1
     return 2 + this._curLine - this._curSplice[0];
   }
 
```

Reply:
Thanks for pointing to `_putCurLineInSplice`. Will take a look.

Imo, the "must end with newline" constraint isn't true. However, I'm not sure if the "documents must end with newline" constraint applies, though... I think as long as mutateAttributionLines keeps working without any change, I can ensure newlines are present on every line.

Reply:
What should happen if an op removes the last line's newline but that op is immediately followed by another op that restores the newline? For example:

```javascript
const text = 'x\n';
const lines = [text];
const cs = '=1|1-1|1+1$\n'; // Keep a char, then remove a newline, then add a newline.
Changeset.mutateTextLines(cs, lines);
assert.equal(lines.join(''), 'x\n');
assert.equal(Changeset.applyToText(cs, text), 'x\n');
```

Similarly, what if an op adds text to the end of the document without a trailing newline, then the next op adds the newline:

```javascript
const text = 'x\n';
const lines = [text];
const cs = '|1=2+1|1+1$y\n'; // Keep existing text, add a char, then add a newline.
Changeset.mutateTextLines(cs, lines);
assert.equal(lines.join(''), 'x\ny\n');
assert.equal(Changeset.applyToText(cs, text), 'x\ny\n');
```

If we force newlines to always be present in each entry in `this._lines` then we're going to end up inserting extra newlines when we don't want them. I think the way to fix this is to strip the newlines from `this._lines` and add them back before returning.